src/gluonts/ev/api.py

```diff
@@ -0,0 +1,155 @@
+    Optional,
+    Protocol,
+    runtime_checkable,
+)
+
+import numpy as np
+
+from gluonts.dataset.split import TestData
+from gluonts.time_feature.seasonality import get_seasonality
+from gluonts.model.predictor import Predictor
+from .stats import seasonal_error
+from .aggregations import Aggregation
+
+
+@runtime_checkable
+class Metric(Protocol):
+    def __call__(self, axis: Optional[int] = None) -> "MetricEvaluator":
+        raise NotImplementedError
+
+
+class EvaluationDataBatch:
+    """Used to add batch dimension
+    Should be replaced by a `ForecastBatch` eventually"""
+
+    def __init__(self, values) -> None:
+        self.values = values
+
+    def __getitem__(self, name):
+        return np.array([self.values[name]])
+
+
+def construct_data(
+    test_data: TestData, predictor: Predictor, **predictor_kwargs
+) -> Iterator[Dict[str, np.ndarray]]:
+    forecasts = predictor.predict(dataset=test_data.input, **predictor_kwargs)
+
+    for input, label, forecast in zip(
+        test_data.input, test_data.label, forecasts
+    ):
+        batching_used = False  # isinstance(forecast, ForecastBatch)
+
+        non_forecast_data = {
+            "label": label["target"],
+            "seasonal_error": seasonal_error(
+                input["target"],
+                seasonality=get_seasonality(freq=forecast.start_date.freqstr),
+            ),
+        }
+        joint_data = ChainMap(non_forecast_data, forecast)
+
+        yield joint_data if batching_used else EvaluationDataBatch(joint_data)
+
+
+class MetricEvaluator:
+    def evaluate(
+        self, test_data: TestData, predictor: Predictor, **predictor_kwargs
+    ) -> np.ndarray:
+        data_batches = construct_data(test_data, predictor, **predictor_kwargs)
+
+        for data in data_batches:
+            self.update(data)
+
+        return self.get()
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        raise NotImplementedError
+
+    def get(self) -> np.ndarray:
+        raise NotImplementedError
+
+
+@dataclass
+class StandardMetricEvaluator(MetricEvaluator):
+    """A "standard metric" consists of a metric function and aggregation
+    strategy."""
+
+    map: Callable
+    aggregate: Aggregation
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        self.aggregate.step(self.map(data))
+
+    def get(self) -> np.ndarray:
+        return self.aggregate.get()
+
+
+@dataclass
+class DerivedMetricEvaluator(MetricEvaluator):
+    """A "derived metric" depends on the prior calculation of "standard
+    metrics"."""
+
+    metrics: Dict[str, StandardMetricEvaluator]
+    post_process: Callable
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        for metric in self.metrics.values():
+            metric.update(data)
+
+    def get(self) -> np.ndarray:
+        return self.post_process(
+            **{name: metric.get() for name, metric in self.metrics.items()}
+        )
+
+
+@dataclass
+class MultiMetricEvaluator:
```

Comment:
This should have a different name, since it's not an `Evaluator` like the others are.

Reply:
It's confusing, indeed. I was thinking about just `MultiMetric` but then, it sounds like it satisfies the `Metric` Protocol...

Or maybe `MetricBatch` or `MetricGroup`?

Reply:
I renamed the "Evaluators" in [this commit](https://github.com/awslabs/gluonts/pull/2267/commits/a381d4aa62c98825d687cd52c1ad1ccb40541056). The "Metric" in `MetricEvaluator` seemed redundant and things got a bit shorter.

MetricEvaluator -> Evaluator
StandardMetricEvaluator -> DirectEvaluator
DerivedMetricEvaluator -> DerivedEvaluator  (not sure if it was better with the "metric" here...)
MultiMetricEvaluator -> MetricGroup

It's still not perfect, I'm open for other ideas.